nixos/modules/services/matrix/synapse.nix

```diff
@@ -719,66 +814,128 @@ in {
+            ExecReload = "${pkgs.util-linux}/bin/kill -HUP $MAINPID";
+            Restart = "on-failure";
+            UMask = "0077";
+
+            # Security Hardening
+            # Refer to systemd.exec(5) for option descriptions.
+            CapabilityBoundingSet = [ "" ];
+            LockPersonality = true;
+            NoNewPrivileges = true;
+            PrivateDevices = true;
+            PrivateTmp = true;
+            PrivateUsers = true;
+            ProcSubset = "pid";
+            ProtectClock = true;
+            ProtectControlGroups = true;
+            ProtectHome = true;
+            ProtectHostname = true;
+            ProtectKernelLogs = true;
+            ProtectKernelModules = true;
+            ProtectKernelTunables = true;
+            ProtectProc = "invisible";
+            ProtectSystem = "strict";
+            ReadWritePaths = [ cfg.dataDir ];
+            RemoveIPC = true;
+            RestrictAddressFamilies = [ "AF_INET" "AF_INET6" "AF_UNIX" ];
+            RestrictNamespaces = true;
+            RestrictRealtime = true;
+            RestrictSUIDSGID = true;
+            SystemCallArchitectures = "native";
+            SystemCallFilter = [ "@system-service" "~@resources" "~@privileged" ];
+          };
+        }
+        // targetConfig;
+        genWorkerService = workerCfg:
+          let
+            name = workerCfg.worker_name;
+            workerConfigFile = format.generate "worker-${name}.yaml" workerCfg;
+          in
+          {
+            "matrix-synapse-worker-${name}" = lib.mkMerge [
+              baseServiceConfig
+              {
+                description = "Synapse Matrix worker ${name}";
+                # make sure the main process starts first for potential database migrations
+                after = [ "matrix-synapse.service" ];
+                serviceConfig = {
+                  ExecStart = ''
+                    ${cfg.package}/bin/synapse_worker \
+                      ${ concatMapStringsSep "\n  " (x: "--config-path ${x} \\") ([ configFile workerConfigFile ] ++ cfg.extraConfigFiles) }
+                      --keys-directory ${cfg.dataDir}
+                  '';
+                };
+              }
+            ];
+          };
+      in
+      lib.mkMerge ([
+        {
+          matrix-synapse = lib.mkMerge [
+            baseServiceConfig
+            {
+              description = "Synapse Matrix homeserver";
+              preStart = ''
+                ${cfg.package}/bin/synapse_homeserver \
+                  --config-path ${configFile} \
+                  --keys-directory ${cfg.dataDir} \
+                  --generate-keys
+              '';
+              serviceConfig = {
+                ExecStartPre = [
+                  ("+" + (pkgs.writeShellScript "matrix-synapse-fix-permissions" ''
+                    chown matrix-synapse:matrix-synapse ${cfg.settings.signing_key_path}
+                    chmod 0600 ${cfg.settings.signing_key_path}
+                  ''))
```

Comment:
Are we reducing the permissions? if so that is not super secure.

Reply:
I haven't changed this part of the module, but I don't really see why we shouldn't chmod it to 0600?

Reply:
This chmod exists because once upon a time the key files were created with too wide permissions.

79e675444caf7b491b2c0d25277b046d3f6d8e04

Reply:
We could deprecate this based on stateVersion but that's something for another PR.